crates/db_views/src/post_view.rs

```diff
@@ -433,10 +476,93 @@ pub struct PostQuery<'a> {
   pub is_profile_view: bool,
   pub page: Option<i64>,
   pub limit: Option<i64>,
+  pub page_after: Option<PaginationCursorData>,
+  pub page_before_or_equal: Option<PaginationCursorData>,
 }
 
 impl<'a> PostQuery<'a> {
   pub async fn list(self, pool: &mut DbPool<'_>) -> Result<Vec<PostView>, Error> {
+    if self.listing_type == Some(ListingType::Subscribed)
+      && self.community_id.is_none()
+      && self.local_user.is_some()
+      && self.page_before_or_equal.is_none()
+    {
+      // first get one page for the most popular community to get an upper bound for the the page end for the real query
+      // the reason this is needed is that when fetching posts for a single community PostgreSQL can optimize
+      // the query to use an index on e.g. (=, >=, >=, >=) and fetch only LIMIT rows
+      // but for the followed-communities query it has to query the index on (IN, >=, >=, >=)
+      // which it currently can't do at all (as of PG 16). see the discussion here:
+      // https://github.com/LemmyNet/lemmy/issues/2877#issuecomment-1673597190
+      //
+      // the results are correct no matter which community we fetch these for, since it basically covers the "worst case" of the whole page consisting of posts from one community
+      // but using the largest community decreases the pagination-frame so make the real query more efficient
+      use lemmy_db_schema::schema::{
+        community_aggregates::dsl::{community_aggregates, community_id, users_active_month},
+        community_follower::dsl::{
+          community_follower,
+          community_id as follower_community_id,
+          person_id,
+        },
+      };
+      let (limit, offset) = limit_and_offset(self.page, self.limit)?;
```

Comment:
You can pry remove, as we should make this a breaking change. 

Add a comment if you would to `limit_and_offset`, that it should be deprecated / removed in favor of cursor-based pagination. Obvi doesn't have to be fully removed for this PR, as comment paging still needs to be worked out.

Reply:
Deprecated definitely, but (same as with other breaking changes) I would prefer to keep it around for one version. It doesn't really cost us anything and if we just instantly remove it it will cause chaos as all existing apps break and all instance admins have to make a decision on whether or not and when to upgrade. So I'm pretty sure if we just instantly break everything that will delay network-wide rollout (and overall quality) by a lot.